src/scheduler/Scheduler.cpp

```diff
@@ -500,71 +454,94 @@ int Scheduler::scheduleFunctionsOnHost(const std::string& host,
-    faabric::Message msgCopy = msg;
-    std::vector<faabric::Message> msgs = { msg };
-    faabric::BatchExecuteRequest req = faabric::util::batchExecFactory(msgs);
+    auto req = faabric::util::batchExecFactory();
+    *req->add_messages() = msg;
 
     // Specify that this is a normal function, not a thread
-    req.set_type(req.FUNCTIONS);
+    req->set_type(req->FUNCTIONS);
 
     // Make the call
     callFunctions(req, forceLocal);
 }
 
-std::vector<unsigned int> Scheduler::getRecordedMessagesAll()
+void Scheduler::clearRecordedMessages()
+{
+    recordedMessagesAll.clear();
+    recordedMessagesLocal.clear();
+    recordedMessagesShared.clear();
+}
+
+std::vector<faabric::Message> Scheduler::getRecordedMessagesAll()
 {
     return recordedMessagesAll;
 }
 
-std::vector<unsigned int> Scheduler::getRecordedMessagesLocal()
+std::vector<faabric::Message> Scheduler::getRecordedMessagesLocal()
 {
     return recordedMessagesLocal;
 }
 
-std::vector<std::pair<std::string, unsigned int>>
+std::vector<std::pair<std::string, faabric::Message>>
 Scheduler::getRecordedMessagesShared()
 {
     return recordedMessagesShared;
 }
 
-void Scheduler::incrementInFlightCount(const faabric::Message& msg)
+void Scheduler::incrementInFlightCount(const faabric::Message& msg, int count)
 {
     const std::string funcStr = faabric::util::funcToString(msg, false);
-    inFlightCounts[funcStr]++;
+
+    inFlightCounts[funcStr] += count;
+
     thisHostResources.set_functionsinflight(
-      thisHostResources.functionsinflight() + 1);
+      thisHostResources.functionsinflight() + count);
 }
 
-void Scheduler::addFaaslets(const faabric::Message& msg)
+void Scheduler::addFaasletsForBatch(const faabric::Message& msg)
 {
     auto logger = faabric::util::getLogger();
+
+    // Check whether we already have a faaslet for this function on this host
     const std::string funcStr = faabric::util::funcToString(msg, false);
 
     // If we've not got one faaslet per in-flight function, add one
     int nFaaslets = faasletCounts[funcStr];
+    if (nFaaslets == 0) {
+        doAddFaaslets(msg, 1);
+    }
+}
+
+void Scheduler::addFaaslets(const faabric::Message& msg)
+{
+    auto logger = faabric::util::getLogger();
+    const std::string funcStr = faabric::util::funcToString(msg, false);
+
+    // If we've not got one faaslet per in-flight function, try to add enough
+    int nFaaslets = faasletCounts[funcStr];
     int inFlightCount = inFlightCounts[funcStr];
-    bool needToScale = nFaaslets < inFlightCount;
 
-    if (needToScale) {
+    int difference = inFlightCount - nFaaslets;
+
+    if (difference > 0) {
         logger->debug(
           "Scaling {} {}->{} faaslets", funcStr, nFaaslets, nFaaslets + 1);
+        doAddFaaslets(msg, difference);
+    }
+}
+
+void Scheduler::doAddFaaslets(const faabric::Message& msg, int count)
+{
+    const std::string funcStr = faabric::util::funcToString(msg, false);
 
-        // Increment faaslet count
```

Comment:
Adding an executor is now a case of creating a new instance of the class, rather than sending a bind message to a pool